lib/kubecfg.libsonnet

```diff
@@ -56,4 +56,55 @@
   // to refer to submatches.  Regex is as implemented in golang regexp
   // package (python-ish).
   regexSubst:: std.native("regexSubst"),
+
+  // isK8sObject(o): Return true iff o is a Kubernetes object.
+  isK8sObject(o):: (
+    std.isObject(o) &&
+    std.objectHas(o, "apiVersion") &&
+    std.objectHas(o, "kind")
+  ),
+
+  // Private helper function for map/fold.
+  local isK8sList(o) = $.isK8sObject(o) && o.apiVersion == "v1" && o.kind == "List",
+
+  // deepMap(func, o): Apply the given function to each Kubernetes
+  // object in nested collection o, preserving the structure of o.
+  deepMap(func, o):: (
+    if isK8sList(o) then
+      o + {items: [func(item) for item in super.items]}
+    else if $.isK8sObject(o) then
+      func(o)
+    else if std.isObject(o) then
+      {[k]: $.deepMap(func, o[k]) for k in std.objectFields(o)}
+    else if std.isArray(o) then
+      [$.deepMap(func, elem) for elem in o]
+    else if o == null then
+      null
+    else
+      error ("o must be an object or array of k8s objects, found " + std.type(o))
+  ),
+
+  // fold(func, o, init): Apply the given function to each Kubernetes
+  // object in nested collection o, accumulating a result as we go.
+  // Function arg is invoked with arguments (accumulator, object).
+  fold(func, o, init):: (
+    if isK8sList(o) then
+      $.fold(func, o.items, init)
+    else if $.isK8sObject(o) then
+      func(init, o)
+    else if std.isObject(o) then
+      $.fold(func, [o[k] for k in std.objectFields(o)], init)
+    else if std.isArray(o) then
+      std.foldl(function (running, elem) $.fold(func, elem, running), o, init)
+    else if o == null then
+      init
+    else
+      error ("o must be an object or array of k8s objects, found " + std.type(o))
+  ),
+
+  // byGvkName(accum, o): Helper for 'fold'.  This accumulates a
+  // two-level collection of objects by 'apiVersion.kind'
+  // (GroupVersionKind) and then object 'name'.
+  byGvkName(accum, o)::
+    accum + {[o.apiVersion + "." + o.kind]+: {[o.metadata.name]: o}},
```

Comment:
one last thing: 

the current byGvk function silently overwrites resources that have the same name:

```
{
  _demo:: [
    { kind: 'Config', apiVersion: 'v1', metadata: { name: 'foo' } },
    { kind: 'Config', apiVersion: 'v1', metadata: { name: 'foo' } },
  ],

  demo: kubecfg.fold(kubecfg.byGvkName, $._demo, {}),
```

produces:

```
{
   "demo": {
      "v1.Config": {
         "foo": {
            "apiVersion": "v1",
            "kind": "Config",
            "metadata": {
               "name": "foo"
            }
         }
      }
   }
}
```

perhaps we should error?

Reply:
something like:

```
byGvkName(accum, o):: accum { [o.apiVersion + '.' + o.kind]+: { assert !(o.metadata.name in super), [o.metadata.name]: o } },
```

Reply:
Or with some better error message:

```
  byGvkName(accum, o):: accum { [o.apiVersion + '.' + o.kind]+: local name = o.metadata.name; { assert !( name in super) : 'duplicate resource name %s' % name, [name]: o } },
```

Reply:
anyway we can improve this in another PR

Reply:
Oh!  This uncovers a good point.  It should probably be by gvk + _namespace_ + name.

Reply:
In particular, I think the above will show up a weakness in my existing renderHelmChart PR.  At the moment I add the namespace to _all_ jsonnet objects, regardless of whether they're _actually_ namespaced or not (k8s will ignore/drop it on apply, but it will look weird in a jsonnet function like proposed above).

Reply:
so how should we proceed here?

Reply:
in my personal experience i do have 2 functions similar to those , one with NS and one without it , and i use them depending on the situation

90% of the time the GVKName is enough since the `yaml` i am importing in jsonnet is all about a single namespace.

Reply:
> so how should we proceed here?

I don't care particularly :)  I just want something that y'all will approve so I can pop my PR stack and eventually merge the helm chart PR.

As mentioned previously, I think these 'layouts' functions are _necessarily_ opinionated and specialised (which is why I'd prefer not to have them in the kubecfg library at all).  If we add them (which I'm also ok with), then we shouldn't be surprised that they don't deal with all possible cases.

The PR now has a `gvkName` and `gvkNsName`.